### ISSUE TITLE
How to reference Microsoft.AspNetCore.*.Analyzers

### ISSUE THREAD

**Issue by ToddThomson** [dotnet/aspnetcore#14675]
AspNetCore analyzers are implicitly referenced in an AspNetCore 3 Web project. How do we reference the analyzers in an AspNetCore 3 Library project ( or Razor library project )?

**Comment by pranavkm** [comment]
We don't have a story for referencing these analyzers from projects that don't target the Web SDK. We could consider moving it the analyzers to the Razor SDK so at least they'd be usable in RCL projects.

@rynowak \ @DamianEdwards what do you think?

**Comment by DamianEdwards** [comment]
Which analyzers in particular are you interested in using from a library that doesn't use the WebSDK?

**Comment by ToddThomson** [comment]
@DamianEdwards I am migrating a project to AspNetCore 3. (Project Sdk="Microsoft.NET.Sdk") I am targeting .netcoreapp3.0 and adding,
```
<ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
  </ItemGroup>
```

Within my .Net Core 3 library I utilize services, Mvc controllers and razor views. I use the library as a module component within an MVC Web project.

I am new to using analyzers for AspNetCore and NetCore. I saw that they were implicitly available int the MVC Web project and thought I could put them to use in the component library project too.

In AspNetCore 2 I was using the embedded file provider to access the embedded views in the component library. I realize that using a Razor Library may be the better approach. However, the Razor Library project type does not include any analyzers.

Since the component library utilizes Mvc Controllers and Razor views, I would like to use all the available AspNetCore.*.Analyzers.

**Comment by rynowak** [comment]
@DamianEdwards - the solution that I'd really like to have here is to make it possible for a FrameworkReference to include analyzers. This is a gap between packages (can depend on analyzers) and FrameworkReference (can't include analyzers).

**Comment by ToddThomson** [comment]
@rynowak Yes, I agree. Before posting this issue, my initial thought was that analyzers might have been added through a FrameworkReference parameter in:
```
<ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
 </ItemGroup>
```
A Microsoft.AspNetCore.*.Analyzers and/Or Microsoft.NetCore.*.Analyzers package(s) would be simple. They allow the developer to choose! I find it a bit frustrating when MS decided that I no longer have a choice ( hopefully that wasn't too negative ).

**Comment by DamianEdwards** [comment]
Thanks, I understand the scenario now. 

As @rynowak says, in 3.0 we include Analyzers as part of the SDK which was part of a larger overall redesign of how various bits come together in .NET Core projects (as an aside, we usually refer to this stuff as ".NET Core composition"). The primary goal was to reduce moving parts in order to make building quicker and producing the product (and using it) less complicated and error-prone. As part of that, there was a desire to implicitly include Analyzers in ASP.NET Core projects and the layering that enabled that was the SDK.

Looks like this is a scenario that has a gap as a result of those changes. @rynowak how do the Analyzers come in via the SDK today? I'm assuming it's just some custom includes in the Web SDK? If so, is there a workaround we could give customers, e.g. a blog of MSBuild XML they could put in their csproj that would include the same analyzers?

**Comment by DamianEdwards** [comment]
OK this works as a workaround:

``` xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <WebSdkAnalyzersPath>$(NetCoreRoot)\$(NETCoreSdkVersion)\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs</WebSdkAnalyzersPath>
    <IncludeOpenAPIAnalyzers>true</IncludeOpenAPIAnalyzers>
  </PropertyGroup>

  <ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
  </ItemGroup>

  <ItemGroup>
    <Analyzer
      Include="$([System.IO.Path]::GetFullPath($(WebSdkAnalyzersPath)..))\Microsoft.AspNetCore.Analyzers.dll"
      Condition="'$(Language)'=='C#'" />
    <Analyzer
      Include="$([System.IO.Path]::GetFullPath($(WebSdkAnalyzersPath)..))\Microsoft.AspNetCore.Mvc.Analyzers.dll"
      Condition="'$(Language)'=='C#'" />
    <Analyzer
      Include="$([System.IO.Path]::GetFullPath($(WebSdkAnalyzersPath)..))\Microsoft.AspNetCore.Mvc.Api.Analyzers.dll"
      Condition="'$(Language)'=='C#' AND '$(IncludeOpenAPIAnalyzers)' == 'true'" />
    <Analyzer
      Include="$([System.IO.Path]::GetFullPath($(WebSdkAnalyzersPath)..))\Microsoft.AspNetCore.Components.Analyzers.dll"
      Condition="'$(Language)'=='C#'" />
  </ItemGroup>

</Project>
```

**Comment by ToddThomson** [comment]
@DamianEdwards Thanks. I just love OSS communication flow!

I added the above XML to my csproj. The analyzers now show up in VS 2019 (latest). Cool! There is a slight difference between a Sdk.Web project however. The individual analyzers do not have a drop down button that when pressed lists the individual "rules".

**Comment by ToddThomson** [comment]
@

**Comment by ToddThomson** [comment]
@DamianEdwards @rynowak I thought I might add:

VS 2019 notified me to use the NuGet package ```Microsoft.CodeAnalysis.FxCopAnalyzers``. I decided that I would give it a try. In general, the analyzer gave me a few insights into my code. This led me to noticing that my Sdk.Web projects had analyzers that provided meaningful build warnings. So I then wanted all my projects utilize analyzers ( I believe they really do provide meaningful build warnings, so I will use them in my prerelease "debug" builds ).

So although I can appreciate that simplifying the addition of analyzers for Sdk.Web projects is a nice objective I would much rather go the NuGet route. It is simple and consistent with how other analyzers are brought into a project. Just my opinion thought.

BTW: Thank-you for finally bringing EF6.3 to .NetCore 3. The final hurdle for me to update my old and tired ASP.NET projects!

**Comment by rynowak** [comment]
> So although I can appreciate that simplifying the addition of analyzers for Sdk.Web projects is a nice objective I would much rather go the NuGet route. It is simple and consistent with how other analyzers are brought into a project. Just my opinion though.

Thanks for your feedback, it's really valuable. 

One question - do you mean to say "I would have preferred nuget packages to this gross workaround" or do you mean to say "I wish these analyzers were required a specific opt-in like the FxCop ones".

We didn't expect many users to be in your position, and what we intended is that they're on by default. The analyzers that are part of ASP.NET Core aren't really opinions or suggestions, they're meant to help you find problems faster 😆

**Comment by ToddThomson** [comment]
@rynowak I have learned over the past few days that Analyzers do work! Since you asked, this is my personal preference:

I feel that the use of analyzers should be up to the developer. They should be available for inclusion in any type of project. This is why I would keep it simple and consistent and use NuGet packages.

That said, I like also like your proposal to add analyzers via the FrameworkReference rather than the through the project sdk property. Analyzers would be enabled by default.

Again, when I realized that AspNetCore analyzers were not part of the library project I first looked for them in NuGet. My second thought was that were somehow tied to the Microsoft.AspCoreNet.App framework reference.

I hope that helps.
